### DISCUSSION TITLE
[A-L] (24 aprile 2024) Esercizio 1

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#57] [Esercitazioni A-L]
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/40034429/6f15431a-ac97-43b7-b5a3-b39f1ef77bd6)

**Comment by Traster-uni** [comment]
Questa è la mia soluzione, a mio parere fin troppo semplicistica, al solito commenti e alternative sono benvenuti.
![WhatsApp Image 2024-06-10 at 8 03 29 PM](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84441861/12672b6b-3233-4a3a-80f8-85aee132e5a3)

**Comment by Elia-Belli** [comment]
## Update MST
Siccome $(u,v)$ non è stato inizialmente scelto per l'inserimento in $T$ vuol dire che avrebbe introdotto un ciclo: analizzando gli archi di tale ciclo possiamo valutare quale di questi sia il più pesante e sostituirlo con $(u,v)$ in modo da ridurre il peso, se non è lui stesso.

```python
def update_MST(T,u,v,new_w):
    P = BFS(T,u) #vettore dei padri da u in T
   
    max = new_w  #peso dell'arco (s,t) di peso maggiore nel ciclo
    s = u
    t = v
    z = v
    while(P[z] != z):
        if(w(z,P[z]) > max):
            max = w(z,P[z])
            s = z
            t = P[z]
        z = P[z]
    
    if(max != new_w):
        T.remove((s,t))
        T.add((u,v))
```

**Comment by daromacarlo** [comment]
![immagine](https://github.com/user-attachments/assets/204ea283-e8f3-4127-b450-73f94d23dedb)
